test/jdk/java/foreign/TestBufferStack.java

```diff
@@ -0,0 +1,281 @@
+                outOfStack = hugeFrame.allocate(4);
+                assertEquals(hugeFrame.scope(), outOfStack.scope());
+                assertTrue(outOfStack.asOverlappingSlice(stackSegment).isEmpty());
+            }
+            assertThrows(IllegalStateException.class, () -> outOfStack.get(JAVA_INT, 0));
+
+            // Outer segments are still accessible.
+            segment11.get(JAVA_INT, 0);
+            segment12.get(JAVA_INT, 0);
+        }
+    }
+
+    @Test
+    void testScopedAllocationVt() {
+        VThreadRunner.run(this::testScopedAllocation);
+    }
+
+    static {
+        System.loadLibrary("TestBufferStack");
+    }
+
+    private static final MemoryLayout HVAPoint3D = structLayout(NativeTestHelper.C_DOUBLE, C_DOUBLE, C_DOUBLE);
+    private static final MemorySegment UPCALL_MH = upcallStub(TestBufferStack.class, "recurse", FunctionDescriptor.of(HVAPoint3D, C_INT));
+    private static final MethodHandle DOWNCALL_MH = downcallHandle("recurse", FunctionDescriptor.of(HVAPoint3D, C_INT, ADDRESS));
+
+    public static MemorySegment recurse(int depth) {
+        try {
+            return (MemorySegment) DOWNCALL_MH.invokeExact((SegmentAllocator) Arena.ofAuto(), depth, UPCALL_MH);
+        } catch (Throwable e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Test
+    void testDeepStack() {
+        // Each downcall and upcall require 48 bytes of stack.
+        // After five allocations we start falling back.
+        MemorySegment point = recurse(10);
+        assertEquals( 12.0, point.getAtIndex(C_DOUBLE, 0));
+        assertEquals(11.0, point.getAtIndex(C_DOUBLE, 1));
+        assertEquals( 10.0, point.getAtIndex(C_DOUBLE, 2));
+    }
+
+    @Test
+    void testDeepStackVt() {
+        VThreadRunner.run(this::testDeepStack);
+    }
+
+    @Test
+    void equals() {
+        var first = BufferStack.of(POOL_SIZE);
+        var second = BufferStack.of(POOL_SIZE);
+        assertNotEquals(first, second);
+        assertEquals(first, first);
+    }
+
+    @Test
+    void allocationSameAsPoolSize() {
+        MemoryLayout twoInts = MemoryLayout.sequenceLayout(2, JAVA_INT);
+        var pool = BufferStack.of(POOL_SIZE);
+        long firstAddress;
+        try (var arena = pool.pushFrame(JAVA_INT)) {
+            var segment = arena.allocate(JAVA_INT);
+            firstAddress = segment.address();
+        }
+        for (int i = 0; i < 10; i++) {
+            try (var arena = pool.pushFrame(twoInts)) {
+                var segment = arena.allocate(JAVA_INT);
+                assertEquals(firstAddress, segment.address());
+                var segmentTwo = arena.allocate(JAVA_INT);
+                assertNotEquals(firstAddress, segmentTwo.address());
+            }
+        }
+    }
+
+    @Test
+    void allocationSameAsPoolSizeVt() {
+        VThreadRunner.run(this::allocationSameAsPoolSize);
+    }
+
+    @Test
+    void allocationCaptureStateLayout() {
```

Comment:
This test is not really testing much re. capture state layout. It feels like the only added thing w.r.t. `allocationSameAsPoolSize` is that it also checks that additional allocation will fail with IOOBE -- maybe these two tests should be merged?